src/resolvelib/resolvers.py

```diff
@@ -435,8 +435,33 @@ def resolve(
                 unsatisfied_names
             )
 
+            if len(unsatisfied_names) > 1:
+                filtered_unstatisfied_names = list(
+                    self._p.narrow_requirement_selection(
+                        identifiers=unsatisfied_names,
+                        resolutions=self.state.mapping,
+                        candidates=IteratorMapping(
+                            self.state.criteria,
+                            operator.attrgetter("candidates"),
+                        ),
+                        information=IteratorMapping(
+                            self.state.criteria,
+                            operator.attrgetter("information"),
+                        ),
+                        backtrack_causes=self.state.backtrack_causes,
+                    )
+                )
+            else:
+                filtered_unstatisfied_names = unsatisfied_names
+
             # Choose the most preferred unpinned criterion to try.
-            name = min(unsatisfied_names, key=self._get_preference)
+            if len(filtered_unstatisfied_names) > 1:
+                name = min(
+                    filtered_unstatisfied_names, key=self._get_preference
+                )
+            else:
+                name = filtered_unstatisfied_names[0]
```

Comment:
why need this if-else? `min` should return immediately if the length is 1

Reply:
`min` still calls then key function if there is 1 element, the point is to avoid an unnecessary call to `get_preference`

Reply:
If that's the logic, it's worth a comment. (Also, I'd argue that `min` shouldn't do that, but I'm sure there's some pathological case out there that would break if it changed 🙁)

Reply:
Updated